### DISCUSSION TITLE
Programmatic navigation API to trigger view transition animation

### DISCUSSION THREAD

**Discussion by yjl9903** [withastro/roadmap#653] [Proposal]
### Body

# Summary

Provide programmatic API for navigation to trigger view transition animation.

# Background & Motivation

Now, only clicking the `a` tag on the page will trigger the view transition. But if I use something like `window.open(...)` in the JavaScript logic, it still refresh the full page.

# Example

Something like the following code snippet. Invoke the `navigate` function will navigate to the corresponding URL and view transition animation also display as usual.

```js
// some.js

import { navigate } from 'astro:transition/client'

navigate('/somewhere'); or
navigate('/somewhere', options);
```

**Comment by timokoessler** [comment]
This would be a very useful feature that I directly missed after trying View Transitions. 👍

**Reply by matthewp** [reply]
We'll definitely do something like this

**Comment by martrapp** [comment]
I will look into this

**Comment by martrapp** [comment]
**I will update this section** to reflect the current discussions.

The current plan for the API is like this:
```
import { navigate } from 'astro:transition/client'

navigate('/somewhere'); or
navigate('/somewhere', options);
```
to some extent similar to https://developer.mozilla.org/en-US/docs/Web/API/Navigation/navigate.

The API does not provide special functions for forward & backward navigation.
It is integrated with the browser's history functions.
You can combine `navigate()` with `history.back()`, `history.forward()`, & `history.go()`

* As of Astro 3.2 the API supports `{history: 'replace' | 'push' | 'auto'}` as option. See Astro Docs for documentation.

**Reply by timokoessler** [reply]
Thanks for implementing this feature 👍 (https://github.com/withastro/astro/pull/8571)

**Comment by Brian-McBride** [comment]
Thanks for this. 

I might create a full on stack view manager for a mobile experience (app dev)

This is something for those who just need a quick dirty `back` function 

```ts
import { navigate } from 'astro:transitions/client';
import type { Options } from 'astro/transitions/router';
import { action, atom } from 'nanostores';

const $previousView = atom<string | null>(null);

export const setPreviousView = action(
  $previousView,
  'setPreviewView',
  (state, view: string | null) => {
    state.set(view);
  },
);

export const skNavigateBack = (skipFallback?: boolean) => {
  const previousView = $previousView.get();
  if (previousView) {
    navigate(previousView, { history: 'replace' });
  } else if (!skipFallback) {
    navigate('/app', { history: 'replace' });
  }
};

export const skNavigate = (href: string, options?: Options) => {
  setPreviousView(window.location.pathname);
  navigate(href, options);
};
```

**Reply by martrapp** [reply]
It's worth noting that view transitions are compatible with browser history. Users can use the browser `back` and `forward` buttons, and these also work with view transitions. 

On the API side, you can combine `navigate()` directly with the browser's `history.back()`, `history.forward()`, and `history.go()` history functions.

**Comment by matthewp** [comment]
This feature was released! If there are bugs please file them in the main repo. If they are no add-ons to this feature, please open a new proposal. Thank you!
